presto-main/src/main/java/com/facebook/presto/operator/scalar/KDistinct.java

```diff
@@ -0,0 +1,114 @@
+import com.facebook.presto.metadata.SqlScalarFunction;
+import com.facebook.presto.spi.function.FunctionKind;
+import com.facebook.presto.spi.function.Signature;
+import com.facebook.presto.spi.function.SqlFunctionVisibility;
+import com.facebook.presto.sql.gen.lambda.BinaryFunctionInterface;
+import com.google.common.collect.ImmutableList;
+import it.unimi.dsi.fastutil.longs.LongArraySet;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.metadata.BuiltInTypeAndFunctionNamespaceManager.DEFAULT_NAMESPACE;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.ArgumentProperty.functionTypeArgumentProperty;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.ArgumentProperty.valueTypeArgumentProperty;
+import static com.facebook.presto.operator.scalar.ScalarFunctionImplementationChoice.NullConvention.RETURN_NULL_ON_NULL;
+import static com.facebook.presto.spi.function.SqlFunctionVisibility.HIDDEN;
+import static com.facebook.presto.util.Reflection.methodHandle;
+
+public final class KDistinct
+        extends SqlScalarFunction
+{
+    public static final KDistinct K_DISTINCT = new KDistinct();
+    private static final MethodHandle STATE_FACTORY = methodHandle(KDistinct.class, "createState");
+
+    private KDistinct()
+    {
+        super(new Signature(
+                QualifiedObjectName.valueOf(DEFAULT_NAMESPACE, "k_distinct"),
+                FunctionKind.SCALAR,
+                ImmutableList.of(),
+                ImmutableList.of(),
+                BOOLEAN.getTypeSignature(),
+                ImmutableList.of(BIGINT.getTypeSignature(), BIGINT.getTypeSignature()),
+                false));
+    }
+
+    @Override
+    public SqlFunctionVisibility getVisibility()
+    {
+        return HIDDEN;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "add a hash (long) if it is not already seen and k elements are not seen";
+    }
+
+    @Override
+    public BuiltInScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, FunctionAndTypeManager functionAndTypeManager)
+    {
+        return new BuiltInScalarFunctionImplementation(
+                false,
+                ImmutableList.of(
+                        valueTypeArgumentProperty(RETURN_NULL_ON_NULL),
+                        functionTypeArgumentProperty(BinaryFunctionInterface.class)),
+                methodHandle(KDistinct.class, "addKey", Object.class, long.class, long.class),
+                Optional.of(methodHandle(KDistinct.class, "createState")));
+    }
+
+    @UsedByGeneratedCode
+    public static boolean addKey(Object state, long hash, long size)
```

Comment:
nit: Why "key"? Perhaps, addHash is clearer.

nit: size is K, right? would it make sense to name it that or use the term limit for clarity? It would be nice to add a comment to this method to explain what it does.

Reply:
yeah I was thinking that. Made it k